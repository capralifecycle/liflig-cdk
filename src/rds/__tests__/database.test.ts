import "@aws-cdk/assert/jest"
import * as ec2 from "aws-cdk-lib/aws-ec2"
import * as rds from "aws-cdk-lib/aws-rds"
import { App, Stack } from "aws-cdk-lib"
import "jest-cdk-snapshot"
import { Database } from ".."

test("create database", () => {
  const app = new App()
  const supportStack = new Stack(app, "SupportStack")
  const stack = new Stack(app, "Stack")

  const vpc = new ec2.Vpc(supportStack, "Vpc")

  const otherSecurityGroup = new ec2.SecurityGroup(
    supportStack,
    "SecurityGroup",
    {
      vpc,
    },
  )

  const database = new Database(stack, "Database", {
    vpc: vpc,
    engine: rds.DatabaseInstanceEngine.postgres({
      version: rds.PostgresEngineVersion.VER_12,
    }),
    instanceType: ec2.InstanceType.of(
      ec2.InstanceClass.BURSTABLE3,
      ec2.InstanceSize.MICRO,
    ),
    instanceIdentifier: "example-database-v1",
    snapshotIdentifier: undefined,
    usePublicSubnets: true,
  })

  database.allowConnectionFrom(otherSecurityGroup)

  expect(stack).toMatchCdkSnapshot()
})

test("should set publiclyAccessible also when in private subnets", () => {
  // This is mainly due to compatibility to previous setups.
  // If the attribute is removed later it will cause a replacement.

  const app = new App()
  const stack = new Stack(app, "Stack")
  const vpc = new ec2.Vpc(stack, "Vpc")

  new Database(stack, "Database", {
    vpc: vpc,
    engine: rds.DatabaseInstanceEngine.postgres({
      version: rds.PostgresEngineVersion.VER_12,
    }),
    instanceType: ec2.InstanceType.of(
      ec2.InstanceClass.BURSTABLE3,
      ec2.InstanceSize.MICRO,
    ),
    instanceIdentifier: "example-database-v1",
    snapshotIdentifier: undefined,
    usePublicSubnets: false,
  })

  expect(stack).toHaveResourceLike("AWS::RDS::DBInstance", {
    PubliclyAccessible: false,
  })
})
