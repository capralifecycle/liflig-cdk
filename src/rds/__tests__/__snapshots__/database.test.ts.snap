// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`create database 1`] = `
Object {
  "Resources": Object {
    "DatabaseF44359DD": Object {
      "DeletionPolicy": "Snapshot",
      "Properties": Object {
        "AllocatedStorage": "25",
        "AllowMajorVersionUpgrade": true,
        "BackupRetentionPeriod": 35,
        "CopyTagsToSnapshot": true,
        "DBInstanceClass": "db.t3.micro",
        "DBInstanceIdentifier": "example-database-v1",
        "DBName": "app",
        "DBSubnetGroupName": Object {
          "Ref": "DatabaseSubnetGroup669F481A",
        },
        "Engine": "postgres",
        "EngineVersion": "12",
        "MasterUserPassword": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{{resolve:secretsmanager:",
              Object {
                "Ref": "DatabaseSecret3B817195",
              },
              ":SecretString:password::}}",
            ],
          ],
        },
        "MasterUsername": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{{resolve:secretsmanager:",
              Object {
                "Ref": "DatabaseSecret3B817195",
              },
              ":SecretString:username::}}",
            ],
          ],
        },
        "MultiAZ": true,
        "PubliclyAccessible": false,
        "StorageEncrypted": true,
        "StorageType": "gp2",
        "VPCSecurityGroups": Array [
          Object {
            "Fn::GetAtt": Array [
              "DatabaseSecurityGroupB7D8C207",
              "GroupId",
            ],
          },
        ],
      },
      "Type": "AWS::RDS::DBInstance",
      "UpdateReplacePolicy": "Snapshot",
    },
    "DatabaseSecret3B817195": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "Description": Object {
          "Fn::Join": Array [
            "",
            Array [
              "Generated by the CDK for stack: ",
              Object {
                "Ref": "AWS::StackName",
              },
            ],
          ],
        },
        "GenerateSecretString": Object {
          "ExcludeCharacters": " %+~\`#$&*()|[]{}:;<>?!'/@\\"\\\\",
          "GenerateStringKey": "password",
          "PasswordLength": 30,
          "SecretStringTemplate": "{\\"username\\":\\"master\\"}",
        },
      },
      "Type": "AWS::SecretsManager::Secret",
      "UpdateReplacePolicy": "Delete",
    },
    "DatabaseSecretAttachmentE5D1B020": Object {
      "Properties": Object {
        "SecretId": Object {
          "Ref": "DatabaseSecret3B817195",
        },
        "TargetId": Object {
          "Ref": "DatabaseF44359DD",
        },
        "TargetType": "AWS::RDS::DBInstance",
      },
      "Type": "AWS::SecretsManager::SecretTargetAttachment",
    },
    "DatabaseSecurityGroupB7D8C207": Object {
      "Properties": Object {
        "GroupDescription": "Security group for Resource database",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "VpcId": Object {
          "Fn::ImportValue": "SupportStack:ExportsOutputRefVpc8378EB38272D6E3A",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "DatabaseSecurityGroupfromSupportStackSecurityGroupF35DA957IndirectPort55C732DA": Object {
      "Properties": Object {
        "Description": "from SupportStackSecurityGroupF35DA957:{IndirectPort}",
        "FromPort": Object {
          "Fn::GetAtt": Array [
            "DatabaseF44359DD",
            "Endpoint.Port",
          ],
        },
        "GroupId": Object {
          "Fn::GetAtt": Array [
            "DatabaseSecurityGroupB7D8C207",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "SourceSecurityGroupId": Object {
          "Fn::ImportValue": "SupportStack:ExportsOutputFnGetAttSecurityGroupDD263621GroupIdDF6F8B09",
        },
        "ToPort": Object {
          "Fn::GetAtt": Array [
            "DatabaseF44359DD",
            "Endpoint.Port",
          ],
        },
      },
      "Type": "AWS::EC2::SecurityGroupIngress",
    },
    "DatabaseSubnetGroup669F481A": Object {
      "Properties": Object {
        "DBSubnetGroupDescription": "Subnet group for Resource database",
        "SubnetIds": Array [
          Object {
            "Fn::ImportValue": "SupportStack:ExportsOutputRefVpcPublicSubnet1Subnet5C2D37C4FFA2B456",
          },
          Object {
            "Fn::ImportValue": "SupportStack:ExportsOutputRefVpcPublicSubnet2Subnet691E08A351552740",
          },
        ],
      },
      "Type": "AWS::RDS::DBSubnetGroup",
    },
  },
}
`;
